addons/account_peppol/models/res_partner.py

```diff
@@ -0,0 +1,71 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import requests
+from lxml import etree
+from hashlib import md5
+from urllib import parse
+
+from odoo import api, fields, models
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    account_peppol_is_endpoint_valid = fields.Boolean(
+        string="PEPPOL endpoint validity",
+        help="The partner's EAS code and PEPPOL endpoint are valid",
+        compute="_compute_account_peppol_is_endpoint_valid",
+        store=True,
+    )
+    account_peppol_validity_last_check = fields.Date(
+        string="Checked on",
+        help="Last Peppol endpoint verification",
+        readonly=True,
+    )
+
+    @api.depends('peppol_eas', 'peppol_endpoint')
+    def _compute_account_peppol_is_endpoint_valid(self):
+        for partner in self:
+            # Every change in peppol_eas or peppol_endpoint should set the validity back to False
+            partner.account_peppol_is_endpoint_valid = False
+
+    def _account_peppol_get_resolvable_smp(self, participant):
+        """
+        This function gives the address for the DNS lookup of the given peppol participant.
+        The SML (Service Metadata Locator) assigns a DNS name to each peppol participant.
+        This DNS name resolves into the SMP (Service Metadata Publisher) of the participant.
+        The DNS address is of the following form:
+        - "http://B-" + hexstring(md5(lowercase(ID-VALUE))) + "." + ID-SCHEME + "." + SML-ZONE-NAME + "/" + url_encoded(ID-SCHEME + "::" + ID-VALUE)
+        (ref:https://peppol.helger.com/public/locale-en_US/menuitem-docs-doc-exchange)
+        :param participant:     peppol participant (ex: '9999:test')
+        """
+        hash_participant = md5(participant.lower().encode()).hexdigest()
+        endpoint_participant = parse.quote_plus(f"iso6523-actorid-upis::{participant}")
+        return f"http://B-{hash_participant}.iso6523-actorid-upis.edelivery.tech.ec.europa.eu/{endpoint_participant}"
+
+    def button_account_peppol_check_partner_endpoint(self):
+        """ A basic check for whether a participant is reachable at the given
+        Peppol participant ID - peppol_eas:peppol_endpoint (ex: '9999:test')
+        """
+        self.ensure_one()
+
+        participant = f'{self.peppol_eas}:{self.peppol_endpoint}'.lower()
+        smp_url = self._account_peppol_get_resolvable_smp(participant)
```

Comment:
I thought it would be cleaner & more readable this way, and more flexible for the future in case we need it somewhere else later, rather than adding one big method. But I can merge these, of course

Reply:
Or move the `requests.get` inside if you want to separate. What we don't want, security-wise, is to have a method that makes a request to an arbitray URL.

Reply:
moved in one :slightly_smiling_face: